File: scripts/notify_pr.py
#!/usr/bin/env python3
"""
PR 通知脚本
用于在 PR 中添加 CI/CD 结果评论
"""

import os
import sys
import json
import argparse


def load_performance_results(results_file):
    """加载性能测试结果"""
    try:
        with open(results_file, 'r') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading performance results: {e}", file=sys.stderr)
        return None


def load_baseline(baseline_file):
    """加载性能基线"""
    try:
        with open(baseline_file, 'r') as f:
            return json.load(f)
    except Exception as e:
        print(f"Error loading baseline: {e}", file=sys.stderr)
        return None


def compare_performance(current, baseline):
    """比较当前性能与基线"""
    results = {
        'regression': False,
        'improvement': False,
        'details': []
    }
    
    if not current or not baseline:
        return results
    
    # 获取 RPS 数据
    current_rps = None
    baseline_rps = None
    
    if 'test_scenarios' in current:
        for scenario in current['test_scenarios']:
            if 'results' in scenario and 'rps' in scenario['results']:
                current_rps = scenario['results']['rps']['value']
                break
    
    if 'baseline' in baseline and 'rps' in baseline['baseline']:
        baseline_rps = baseline['baseline']['rps']
    
    if current_rps and baseline_rps:
        change = ((current_rps - baseline_rps) / baseline_rps) * 100
        
        results['details'].append({
            'metric': 'RPS',
            'baseline': baseline_rps,
            'current': current_rps,
            'change': change
        })
        
        # 回归阈值：RPS 下降超过 10%
        if change < -10:
            results['regression'] = True
        # 改进阈值：RPS 提升超过 10%
        elif change > 10:
            results['improvement'] = True
    
    return results


def generate_pr_comment(workflow_name, conclusion, run_id, run_number, 
                       performance_comparison=None, test_results=None):
    """生成 PR 评论内容"""
    comment = f"## 🤖 CI/CD PR Validation Results\n\n"
    comment += f"**Workflow**: {workflow_name}\n"
    comment += f"**Run**: #{run_number}\n"
    
    # 状态图标
    if conclusion == 'success':
        comment += f"**Status**: ✅ Passed\n\n"
    else:
        comment += f"**Status**: ❌ Failed\n\n"
    
    # 性能比较
    if performance_comparison:
        comment += "### 📊 Performance Comparison\n\n"
        
        if performance_comparison['regression']:
            comment += "⚠️ **Performance Regression Detected**\n\n"
        elif performance_comparison['improvement']:
            comment += "✨ **Performance Improvement Detected**\n\n"
        else:
            comment += "✅ **Performance Stable**\n\n"
        
        comment += "| Metric | Baseline | Current | Change |\n"
        comment += "|--------|----------|---------|--------|\n"
        
        for detail in performance_comparison['details']:
            change_pct = detail['change']
            change_str = f"{change_pct:+.2f}%"
            
            if change_pct < -10:
                change_str = f"🔴 {change_str}"
            elif change_pct > 10:
                change_str = f"🟢 {change_str}"
            
            comment += f"| {detail['metric']} | {detail['baseline']:.2f} | {detail['current']:.2f} | {change_str} |\n"
        
        comment += "\n"
    
    # 测试结果
    if test_results:
        comment += "### 🧪 Test Results\n\n"
        comment += f"- Total: {test_results.get('total', 0)}\n"
        comment += f"- Passed: {test_results.get('passed', 0)} ✅\n"
        comment += f"- Failed: {test_results.get('failed', 0)} ❌\n"
        comment += f"- Skipped: {test_results.get('skipped', 0)} ⏭️\n\n"
    
    # 详细链接
    comment += "### 🔗 Links\n\n"
    comment += f"- [View Workflow Run](https://github.com/adam-ikari/uvhttp/actions/runs/{run_id})\n"
    
    if conclusion == 'success':
        comment += "\nReady for review! 🎉"
    else:
        comment += "\nPlease fix the issues and push a new commit."
    
    return comment


def main():
    parser = argparse.ArgumentParser(description='Generate PR notification comment')
    parser.add_argument('--workflow-name', required=True, help='Workflow name')
    parser.add_argument('--conclusion', required=True, choices=['success', 'failure', 'cancelled', 'skipped'], help='Workflow conclusion')
    parser.add_argument('--run-id', required=True, help='Workflow run ID')
    parser.add_argument('--run-number', required=True, help='Workflow run number')
    parser.add_argument('--performance-results', help='Path to performance results JSON')
    parser.add_argument('--baseline', help='Path to baseline JSON')
    parser.add_argument('--test-results', help='Path to test results JSON')
    parser.add_argument('--output', '-o', help='Output file (default: stdout)')
    
    args = parser.parse_args()
    
    # 加载数据
    performance_data = None
    baseline_data = None
    test_data = None
    
    if args.performance_results:
        performance_data = load_performance_results(args.performance_results)
    
    if args.baseline:
        baseline_data = load_baseline(args.baseline)
    
    if args.test_results:
        with open(args.test_results, 'r') as f:
            test_data = json.load(f)
    
    # 比较性能
    performance_comparison = None
    if performance_data and baseline_data:
        performance_comparison = compare_performance(performance_data, baseline_data)
    
    # 生成评论
    comment = generate_pr_comment(
        args.workflow_name,
        args.conclusion,
        args.run_id,
        args.run_number,
        performance_comparison,
        test_data
    )
    
    # 输出
    if args.output:
        with open(args.output, 'w') as f:
            f.write(comment)
        print(f"Comment written to {args.output}")
    else:
        print(comment)


if __name__ == '__main__':
    main()

File: scripts/ci/notify_pr.js
#!/usr/bin/env node
/**
 * PR notification script
 * 
 * This script generates PR notification comments with CI/CD results,
 * performance comparisons, and test results
 */

const fs = require('fs').promises;

/**
 * Load performance test results
 * @param {string} resultsFile - Path to results file
 * @returns {Promise<Object|null>} Parsed results or null
 */
async function loadPerformanceResults(resultsFile) {
  try {
    const content = await fs.readFile(resultsFile, 'utf-8');
    return JSON.parse(content);
  } catch (error) {
    console.error(`Error loading performance results: ${error.message}`);
    return null;
  }
}

/**
 * Load performance baseline
 * @param {string} baselineFile - Path to baseline file
 * @returns {Promise<Object|null>} Parsed baseline or null
 */
async function loadBaseline(baselineFile) {
  try {
    const content = await fs.readFile(baselineFile, 'utf-8');
    return JSON.parse(content);
  } catch (error) {
    console.error(`Error loading baseline: ${error.message}`);
    return null;
  }
}

/**
 * Compare current performance with baseline
 * @param {Object} current - Current performance data
 * @param {Object} baseline - Baseline performance data
 * @returns {Object} Comparison results
 */
function comparePerformance(current, baseline) {
  const results = {
    regression: false,
    improvement: false,
    details: []
  };
  
  if (!current || !baseline) {
    return results;
  }
  
  // Get RPS data
  let currentRps = null;
  let baselineRps = null;
  
  if (current.test_scenarios) {
    for (const scenario of current.test_scenarios) {
      if (scenario.results && scenario.results.rps) {
        currentRps = scenario.results.rps.value;
        break;
      }
    }
  }
  
  if (baseline.baseline && baseline.baseline.rps) {
    baselineRps = baseline.baseline.rps;
  }
  
  if (currentRps && baselineRps) {
    const change = ((currentRps - baselineRps) / baselineRps) * 100;
    
    results.details.push({
      metric: 'RPS',
      baseline: baselineRps,
      current: currentRps,
      change: change
    });
    
    // Regression threshold: RPS decreases by more than 10%
    if (change < -10) {
      results.regression = true;
    }
    // Improvement threshold: RPS increases by more than 10%
    else if (change > 10) {
      results.improvement = true;
    }
  }
  
  return results;
}

/**
 * Generate PR comment content
 * @param {string} workflowName - Workflow name
 * @param {string} conclusion - Workflow conclusion
 * @param {string} runId - Workflow run ID
 * @param {string} runNumber - Workflow run number
 * @param {Object} performanceComparison - Performance comparison results
 * @param {Object} testResults - Test results
 * @returns {string} PR comment content
 */
function generatePRComment(workflowName, conclusion, runId, runNumber, performanceComparison, testResults) {
  let comment = `## 🤖 CI/CD PR Validation Results\n\n`;
  comment += `**Workflow**: ${workflowName}\n`;
  comment += `**Run**: #${runNumber}\n`;
  
  // Status icon
  if (conclusion === 'success') {
    comment += `**Status**:  Passed\n\n`;
  } else {
    comment += `**Status**:  Failed\n\n`;
  }
  
  // Performance comparison
  if (performanceComparison) {
    comment += `###  Performance Comparison\n\n`;
    
    if (performanceComparison.regression) {
      comment += ` **Performance Regression Detected**\n\n`;
    } else if (performanceComparison.improvement) {
      comment += `✨ **Performance Improvement Detected**\n\n`;
    } else {
      comment += ` **Performance Stable**\n\n`;
    }
    
    comment += `| Metric | Baseline | Current | Change |\n`;
    comment += `|--------|----------|---------|--------|\n`;
    
    for (const detail of performanceComparison.details) {
      const changePct = detail.change;
      let changeStr = `${changePct >= 0 ? '+' : ''}${changePct.toFixed(2)}%`;
      
      if (changePct < -10) {
        changeStr = `🔴 ${changeStr}`;
      } else if (changePct > 10) {
        changeStr = `🟢 ${changeStr}`;
      }
      
      comment += `| ${detail.metric} | ${detail.baseline.toFixed(2)} | ${detail.current.toFixed(2)} | ${changeStr} |\n`;
    }
    
    comment += `\n`;
  }
  
  // Test results
  if (testResults) {
    comment += `### 🧪 Test Results\n\n`;
    comment += `- Total: ${testResults.total || 0}\n`;
    comment += `- Passed: ${testResults.passed || 0} \n`;
    comment += `- Failed: ${testResults.failed || 0} \n`;
    comment += `- Skipped: ${testResults.skipped || 0} ⏭️\n\n`;
  }
  
  // Detailed links
  comment += `### 🔗 Links\n\n`;
  comment += `- [View Workflow Run](https://github.com/adam-ikari/uvhttp/actions/runs/${runId})\n`;
  
  if (conclusion === 'success') {
    comment += `\nReady for review! 🎉`;
  } else {
    comment += `\nPlease fix the issues and push a new commit.`;
  }
  
  return comment;
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  // Parse arguments
  const options = {
    workflowName: null,
    conclusion: null,
    runId: null,
    runNumber: null,
    performanceResults: null,
    baseline: null,
    testResults: null,
    output: null
  };
  
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--workflow-name':
        options.workflowName = args[++i];
        break;
      case '--conclusion':
        options.conclusion = args[++i];
        break;
      case '--run-id':
        options.runId = args[++i];
        break;
      case '--run-number':
        options.runNumber = args[++i];
        break;
      case '--performance-results':
        options.performanceResults = args[++i];
        break;
      case '--baseline':
        options.baseline = args[++i];
        break;
      case '--test-results':
        options.testResults = args[++i];
        break;
      case '--output':
      case '-o':
        options.output = args[++i];
        break;
      default:
        console.error(`Unknown option: ${args[i]}`);
        process.exit(1);
    }
  }
  
  // Validate required arguments
  if (!options.workflowName || !options.conclusion || !options.runId || !options.runNumber) {
    console.error('Error: --workflow-name, --conclusion, --run-id, and --run-number are required');
    console.error('');
    console.error('Usage: node notify_pr.js --workflow-name <name> --conclusion <status>');
    console.error('                         --run-id <id> --run-number <num> [options]');
    console.error('');
    console.error('Options:');
    console.error('  --workflow-name <name>      Workflow name (required)');
    console.error('  --conclusion <status>       Workflow conclusion: success, failure, cancelled, skipped (required)');
    console.error('  --run-id <id>               Workflow run ID (required)');
    console.error('  --run-number <num>          Workflow run number (required)');
    console.error('  --performance-results <path> Path to performance results JSON');
    console.error('  --baseline <path>           Path to baseline JSON');
    console.error('  --test-results <path>       Path to test results JSON');
    console.error('  --output, -o <path>         Output file (default: stdout)');
    console.error('');
    console.error('Example:');
    console.error('  node notify_pr.js --workflow-name "CI" --conclusion success --run-id 123456 --run-number 42');
    process.exit(1);
  }
  
  // Validate conclusion
  const validConclusions = ['success', 'failure', 'cancelled', 'skipped'];
  if (!validConclusions.includes(options.conclusion)) {
    console.error(`Error: Invalid conclusion "${options.conclusion}"`);
    console.error(`Valid values: ${validConclusions.join(', ')}`);
    process.exit(1);
  }
  
  // Load data
  let performanceData = null;
  let baselineData = null;
  let testData = null;
  
  if (options.performanceResults) {
    performanceData = await loadPerformanceResults(options.performanceResults);
  }
  
  if (options.baseline) {
    baselineData = await loadBaseline(options.baseline);
  }
  
  if (options.testResults) {
    try {
      const content = await fs.readFile(options.testResults, 'utf-8');
      testData = JSON.parse(content);
    } catch (error) {
      console.error(`Error loading test results: ${error.message}`);
    }
  }
  
  // Compare performance
  let performanceComparison = null;
  if (performanceData && baselineData) {
    performanceComparison = comparePerformance(performanceData, baselineData);
  }
  
  // Generate comment
  const comment = generatePRComment(
    options.workflowName,
    options.conclusion,
    options.runId,
    options.runNumber,
    performanceComparison,
    testData
  );
  
  // Output
  if (options.output) {
    await fs.writeFile(options.output, comment, 'utf-8');
    console.log(`Comment written to ${options.output}`);
  } else {
    console.log(comment);
  }
}

main().catch(error => {
  console.error('Error:', error.message);
  process.exit(1);
});